### PR TITLE
Portworx refactored, Added Failover test and Snapshot Test

### PR DESCRIPTION
**Pull request by NathanielRose**
- [x] Installation script for strimzi interoperability with Portworx
- [x] Documentation on Portworx setup
- [x] Testing for failover *(Stable)*
- [x] Testing for back up/ snapshot *(Unstable)*

To Do:
- [ ] Change Naming convention to `kcluster`
- [ ] Target 3dsnaps for productionize back up scenario
- [ ] Integrate with Fabrikate Definition (Pending Px helm chart)

### CODE REVIEW

**Comment by NathanielRose** [comment]
@andrebriggs @mtarng @sayar @edaena 
- Im unable to add reviewers, so adding them in this comment.

File: portworx-manual/Portworx.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Using Virtualized Storage with Portworx
```
Comment (andrebriggs) [inline portworx-manual/Portworx.md:1]: @NathanielRose can you rename this file to README.md?

File: portworx-manual/Portworx.md [hunk 2]
```diff
@@ -12,7 +12,7 @@ Portworx includes:
 
 ## Setting up Portworx Manually
 
-To configure your Portworx Virtualization layer with Strimzi, use the following steps:
+To configure your Portworx Virtualization layer with Strimzi, use the following steps or run the [`strimzi-px-install.sh`](strimzi-px-install.sh) script.:
 
 ```
 # Create a secret to give Portworx access to Azure APIs
```
Comment (andrebriggs) [inline portworx-manual/Portworx.md]: Not part of this PR but it would be nice to provide the `az cli` commands to do this or provide a script to generate. Perhaps something like [this](https://github.com/andrebriggs/azure-tf-k8s#create-an-azure-service-principal)

File: portworx-manual/strimzi-px-install.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                                                      --from-literal=AZURE_CLIENT_SECRET="${AZURE_CLIENT_SECRET}"
+
+
+# Generate custom specs for your portworx config. By default, the script uses Premium volume types, 150 GB, Auto Data and 
+# Management network interfaces with Stork, GUI enabled. To customize this config use the api URL to download a custom yaml
+# [https://docs.portworx.com/portworx-install-with-kubernetes/cloud/azure/aks/2-deploy-px/#]
+
+kubectl apply -f temp-px-deploy/px-gen-spec.yaml
+
+until kubectl get pods --all-namespaces | grep -E "kube-system(\s){3}portworx.*1\/1\s*Running+"
```
Comment (andrebriggs) [inline portworx-manual/strimzi-px-install.sh:28]: Perhaps add a comment to say you want to make sure all the Portworx pods in the Daemonset are running

File: portworx-manual/Portworx.md [hunk 4]
```diff
@@ -40,23 +40,40 @@ You can use the `pxctl` command Portworx provides to obtain metrics on how the v
...
+        10.240.0.6      ce4d4aca-fb23-4ac0-ad36-16fa627f1460    aks-nodepool1-40021484-0        Yes             9.8 GiB 147 GiB         Online  Up (This node)  2.1.1.0-6de97a6 4.15.0-1050-azure       Ubuntu 16.04.6 LTS
+        10.240.0.5      1628bbf7-8780-4d52-9a5b-879f0da3fb14    aks-nodepool1-40021484-1        Yes             9.8 GiB 147 GiB         Online  Up              2.1.1.0-6de97a6 4.15.0-1050-azure       Ubuntu 16.04.6 LTS
+        10.240.0.4      001c6978-7ec9-4c79-b5ca-224c3263738a    aks-nodepool1-40021484-2        Yes             9.8 GiB 147 GiB         Online  Up              2.1.1.0-6de97a6 4.15.0-1050-azure       Ubuntu 16.04.6 LTS
+Global Storage Pool
         Total Used      :  29 GiB
+        Total Capacity  :  441 GiB
 ```
 
 3. **Once your storage class definition is applied and you provision PVCs from your application such as Strimzi**, next grab the Portworx volume list which will give details on the provisioned volumes deployed. In this Kafka example we define the size and replica amount of the volumes in our statefulset yaml for Kafka and zookeeper.
```
Comment (andrebriggs) [inline portworx-manual/Portworx.md]: When we enable volume encryption should we expect to see the "encryption" column from the below command be "YES?

`kubectl exec <portworx_pod_name> -n kube-system -- /opt/pwx/bin/pxctl volume list`

Comment (andrebriggs) [reply portworx-manual/Portworx.md]: Also what is the purpose of Step 4? Just to demonstrate that you can look at the detail of a volume?

File: portworx-manual/Portworx.md [hunk 5]
```diff
@@ -110,10 +127,65 @@ Volume  :  924341013124639820
...
-For details on how to snapshot volumes using stork check out - https://portworx.com/run-ha-kafka-azure-kubernetes-service/.
 
 ## Monitoring Portworx
 
-In the provided Portworx spec file, `px-gen-spec.yaml`, there is an added port opening configuration under the Daemonset section allowing Prometheus to scrape Portworx pods for metrics. In addition, added annotations for promethes to access the metric probes have been embedded for Prometheus integration. Navigate to prometheus and view the `px` metrics.
+In the provided Portworx spec file, `px-gen-spec.yaml`, there is an added port opening configuration under the daemonset section allowing Prometheus to scrape Portworx pods for metrics. In addition, added annotations for Prometheus to access the metric probes have been embedded for Prometheus integration. Navigate to prometheus and view the `px` metrics.
 
-> Run `kubectl port-forward -n prometheus <prometheus_server_pod> 9090`
\ No newline at end of file
+> Run `kubectl port-forward -n prometheus <prometheus_server_pod> 9090`
```
Comment (andrebriggs) [inline portworx-manual/Portworx.md]: Perhaps add a screenshot? 
image

File: portworx-manual/Portworx.md [hunk 6]
```diff
@@ -110,10 +127,65 @@ Volume  :  924341013124639820
...
+
+## High Availability & Snapshots with Portworx
+
+High Availability and Disaster Recovery are built into Portworx out of the box. Here are some tips and 
+
+### Portworx Failovers
+
+You can simulate the failover for Portworx by cordoning off one of the nodes and deleting the Kafka Pod deployed on it. To start, follow the steps belor or run the [`px-failovertest.sh`](temp-px-perf/px-failovertest.sh). When the new pod is created it has the same data as the original Pod.
```
Comment (andrebriggs) [inline portworx-manual/Portworx.md]: Perhaps [explain](https://kubernetes.io/docs/concepts/architecture/nodes/#manual-node-administration) that `kubectl cordon` will prevent new pods from being scheduled on the node

Comment (andrebriggs) [inline portworx-manual/Portworx.md]: Spelling `below`

**Comment by andrebriggs** [comment]
One other thing. After following all the instructions I see
```
$ kubectl get pods -n kafka
NAME                                         READY     STATUS    RESTARTS   AGE
kafkaclient-0                                1/1       Running   0          23m
kafkaclient-1                                1/1       Running   0          21m
kafkaclient-2                                1/1       Running   0          21m
kcluster-entity-operator-68c595b8-7vlgj      3/3       Running   0          1h
kcluster-kafka-0                             2/2       Running   0          1h
kcluster-kafka-1                             2/2       Running   0          1h
kcluster-kafka-2                             2/2       Running   0          1h
kcluster-zookeeper-0                         2/2       Running   0          1h
kcluster-zookeeper-1                         2/2       Running   0          1h
kcluster-zookeeper-2                         2/2       Running   0          1h
px-cluster-entity-operator-8dcbb4576-mbs28   3/3       Running   1          22m
px-cluster-kafka-0                           2/2       Running   0          23m
px-cluster-kafka-1                           2/2       Running   0          23m
px-cluster-kafka-2                           2/2       Running   0          23m
px-cluster-zookeeper-0                       2/2       Running   0          24m
px-cluster-zookeeper-1                       2/2       Running   0          24m
px-cluster-zookeeper-2                       2/2       Running   0          24m
strimzi-cluster-operator-68575878f6-x79nh    1/1       Running   0          1h
```

<strike>I have 3 kafka clients. Why?</strike> I see that install script created these. Ideally we wouldn't need 3 clients, right? Ideally we would go through a service with an external ip, right?

Overall this looks good for what we can do while waiting for the helm chart support. Interested to see your response to some of the comments

File: portworx-manual/temp-px-perf/px-snapshottest.sh [hunk 7]
```diff
@@ -0,0 +1,43 @@
+#! /usr/bin/env bash
+
```
Comment (andrebriggs) [inline portworx-manual/temp-px-perf/px-snapshottest.sh]: These empty lines seem to be causing errors when running this script. Should we not use `sh` to run ? the file says `#! /usr/bin/env bash` and ends in `.sh` What I get:

```
$ sh px-snapshottest.sh 
: command not found line 2: 
: command not found line 4: 
: command not found line 6: 
volumesnapshot.volumesnapshot.external-storage.k8s.io "px-kafka-snapshot" created
: command not found line 8: 
Verifying volume snapshot creation
: command not found line 10: 
error: the server doesn't have a resource type "volumesnapshot\r"
: command not found line 12: 
NAME                                                       AGE
k8s-volume-snapshot-38515a1a-b993-11e9-86cf-8ae310e7e8ae   2s
: command not found line 14: 
Deleting kafka broker stateful set and PVC
: command not found line 16: 
Error from server (NotFound): statefulsets.apps "px-cluster-kafka\r" not found
Error from server (NotFound): persistentvolumeclaims "data-px-cluster-kafka-0\r" not found
Error from server (NotFound): persistentvolumeclaims "data-px-cluster-kafka-1\r" not found
Error from server (NotFound): persistentvolumeclaims "data-px-cluster-kafka-2\r" not found
No resources found.
No resources found.
No resources found.
: command not found line 24: 
 Creating new PVC definition from the snapshot.
: command not found line 26: 
persistentvolumeclaim "data-px-cluster-kafka-0" created
No resources found.
: command not found line 29: 
Deploying new stateful set backend by the restored snapshot PVC will be automatically handled by the Strimzi Operator
: command not found line 31: 
: command not found line 33: 
....
```

**Review by andrebriggs — CHANGES_REQUESTED** [review]

File: portworx-manual/temp-px-perf/px-failovertest.sh [hunk 8]
```diff
@@ -0,0 +1,70 @@
+#! /usr/bin/env bash
```
Comment (andrebriggs) [inline portworx-manual/temp-px-perf/px-failovertest.sh:1]: This test fails. @NathanielRose you're checking in a lot in this PR. How about we remove these tests from the PR and you work on them in the background? What I'm seeing:

```
$ sh px-failovertest.sh 
7dba7f60-2613-4cc0-ab49-660ed7179f25
Test Topic: topic-7dba7f60-2613-4cc0-ab49-660ed7179f25
OpenJDK 64-Bit Server VM warning: If the number of processors is expected to increase from one, then you should configure the number of parallel GC threads appropriately using -XX:ParallelGCThreads=N
Created topic topic-7dba7f60-2613-4cc0-ab49-660ed7179f25.
Creating Input Message file.
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 26: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
cat: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 32: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-input-messages.txt: No such file or directory
px-failovertest.sh: line 36: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-output-messages.txt: No such file or directory
px-failovertest.sh: line 40: kill: (54493) - No such process
node "aks-agentpool-11920454-4" cordoned
Cordoning Node: aks-agentpool-11920454-4
NAME                       STATUS                     ROLES     AGE       VERSION   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
aks-agentpool-11920454-0   Ready                      agent     3h        v1.12.8   <none>        Ubuntu 16.04.6 LTS   4.15.0-1050-azure   docker://3.0.6
aks-agentpool-11920454-1   Ready                      agent     3h        v1.12.8   <none>        Ubuntu 16.04.6 LTS   4.15.0-1050-azure   docker://3.0.6
aks-agentpool-11920454-2   Ready                      agent     3h        v1.12.8   <none>        Ubuntu 16.04.6 LTS   4.15.0-1050-azure   docker://3.0.6
aks-agentpool-11920454-3   Ready                      agent     3h        v1.12.8   <none>        Ubuntu 16.04.6 LTS   4.15.0-1050-azure   docker://3.0.6
aks-agentpool-11920454-4   Ready,SchedulingDisabled   agent     3h        v1.12.8   <none>        Ubuntu 16.04.6 LTS   4.15.0-1050-azure   docker://3.0.6

Deleting kafka broker pod: px-cluster-kafka-0
pod "px-cluster-kafka-0" deleted

Deleted kafka broker is now being created on new shared node
NAME                                         READY     STATUS        RESTARTS   AGE       IP            NODE
kafkaclient-0                                1/1       Running       0          57m       10.244.2.9    aks-agentpool-11920454-1
kafkaclient-1                                1/1       Running       0          55m       10.244.3.11   aks-agentpool-11920454-0
kafkaclient-2                                1/1       Running       0          55m       10.244.1.13   aks-agentpool-11920454-4
kcluster-entity-operator-68c595b8-7vlgj      3/3       Running       0          1h        10.244.1.8    aks-agentpool-11920454-4
kcluster-kafka-0                             2/2       Running       0          1h        10.244.4.6    aks-agentpool-11920454-2
kcluster-kafka-1                             2/2       Running       0          1h        10.244.1.7    aks-agentpool-11920454-4
kcluster-kafka-2                             2/2       Running       0          1h        10.244.0.11   aks-agentpool-11920454-3
kcluster-zookeeper-0                         2/2       Running       0          2h        10.244.0.10   aks-agentpool-11920454-3
kcluster-zookeeper-1                         2/2       Running       0          2h        10.244.4.5    aks-agentpool-11920454-2
kcluster-zookeeper-2                         2/2       Running       0          2h        10.244.1.6    aks-agentpool-11920454-4
px-cluster-entity-operator-8dcbb4576-mbs28   3/3       Running       1          56m       10.244.2.10   aks-agentpool-11920454-1
px-cluster-kafka-0                           2/2       Terminating   0          57m       10.244.1.12   aks-agentpool-11920454-4
px-cluster-kafka-1                           2/2       Running       0          57m       10.244.0.13   aks-agentpool-11920454-3
px-cluster-kafka-2                           2/2       Running       0          57m       10.244.3.10   aks-agentpool-11920454-0
px-cluster-zookeeper-0                       2/2       Running       0          58m       10.244.1.11   aks-agentpool-11920454-4
px-cluster-zookeeper-1                       2/2       Running       0          58m       10.244.0.12   aks-agentpool-11920454-3
px-cluster-zookeeper-2                       2/2       Running       0          58m       10.244.3.9    aks-agentpool-11920454-0
strimzi-cluster-operator-68575878f6-x79nh    1/1       Running       0          2h        10.244.1.5    aks-agentpool-11920454-4
Kafka broker messages after cordon are written at:  ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-cordon-output-messages.txt
px-failovertest.sh: line 62: ./temp/topic-7dba7f60-2613-4cc0-ab49-660ed7179f25-cordon-output-messages.txt: No such file or directory
px-failovertest.sh: line 65: kill: (54505) - No such process

Uncordoning the node
node "aks-agentpool-11920454-4" uncordoned
```

**Review by andrebriggs — CHANGES_REQUESTED** [review]

File: portworx-manual/temp-px-deploy/kafka-tls-px.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
+---
+apiVersion: kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: px-cluster
```
Comment (andrebriggs) [inline portworx-manual/temp-px-deploy/kafka-tls-px.yaml:4]: Missing `namespace: kafka`

**Review by andrebriggs — CHANGES_REQUESTED** [review]

File: portworx-manual/Portworx.md (hunk 4 above)
Comment (NathanielRose) [reply portworx-manual/Portworx.md]: Yes, Encryption will simply display `Yes` in the output. I can see this being another unit test.
Step 4 allows us to view the volume respective to the node. We can see the bytes used and capacity to each volume. We're also able to view the associated broker consuming the volume.

File: portworx-manual/Portworx.md (hunk 2 above)
Comment (NathanielRose) [reply portworx-manual/Portworx.md]: I may not be following, wouldn't the infrastructure instructions be based on Bedrock template documentation?

File: portworx-manual/temp-px-perf/px-snapshottest.sh (hunk 7 above)
Comment (NathanielRose) [reply portworx-manual/temp-px-perf/px-snapshottest.sh]: Ran a `dos2unix` on the file. It had trailing windows `/r` chars

File: portworx-manual/temp-px-perf/px-failovertest.sh (hunk 8 above)
Comment (NathanielRose) [reply portworx-manual/temp-px-perf/px-failovertest.sh:1]: `px-failovertest.sh` should work with the latest commit that targets the appropriate broker service ports. Can you verify?

File: portworx-manual/temp-px-deploy/kafka-tls-px.yaml (hunk 9 above)
Comment (NathanielRose) [reply portworx-manual/temp-px-deploy/kafka-tls-px.yaml:4]: The [script](https://github.com/sayar/fabrikate-kafka-sample/pull/14/files/823415cc0f0f8fbd35ecf20e7efb2084f0b48a28#diff-7387bdeabf9b7c8e9e0b5ad4ee69fd85R37) adds the appropriate namespace. Still, add it here?

**Review by NathanielRose — COMMENTED** [review]
Looking for new review

File: portworx-manual/Portworx.md (hunk 2 above)
Comment (andrebriggs) [reply portworx-manual/Portworx.md]: I was looking at this as standalone instructions. I see what you mean though. In isolation it's just anything thing to figure out.

**Comment by andrebriggs** [comment]
@NathanielRose I don't have time to re-run all the tests with this setup. If you say they are working I will trust that. Keep in mind this is why I recommended splitting that up into multiple PRs. Smaller PRs would be appreciated when doing the minimal to integrate for an end to end. I will approve and merge this then we can sort things out.
